File: src/main/java/com/mcommings/campaigner/modules/locations/entities/Country.java
package com.mcommings.campaigner.modules.locations.entities;

import com.mcommings.campaigner.modules.common.entities.Campaign;
import com.mcommings.campaigner.modules.common.entities.Government;
import jakarta.persistence.*;
import lombok.*;

@Getter
@Setter
@Builder
@Entity
@Table(name = "countries",
        uniqueConstraints = {
                @UniqueConstraint(
                        columnNames = {"name", "fk_campaign_uuid"}
                )
        })
@NoArgsConstructor
@AllArgsConstructor
public class Country {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    @Column(nullable = false, unique = true)
    private String name;

    private String description;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_campaign_uuid", nullable = false)
    private Campaign campaign;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_continent")
    private Continent continent;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_government")
    private Government government;
}

File: src/main/frontend/lib/models/location/country.dart
class Country {
  final int id;
  final String name;
  final String description;
  final String fkCampaignUuid;
  final int fkContinent;
  final int fkGovernment;

  const Country({
    required this.id,
    required this.name,
    required this.description,
    required this.fkCampaignUuid,
    required this.fkContinent,
    required this.fkGovernment,
  });

  factory Country.fromJson(Map<String, dynamic> json) {
    return switch (json) {
      {
        'id': int id,
        'name': String name,
        'description': String description,
        'fk_campaign_uuid': String fkCampaignUuid,
        'fk_continent': int fkContinent,
        'fk_government': int fkGovernment,
      } =>
        Country(
          id: id,
          name: name,
          description: description,
          fkCampaignUuid: fkCampaignUuid,
          fkContinent: fkContinent,
          fkGovernment: fkGovernment,
        ),
      _ => throw const FormatException("Failed to load Country."),
    };
  }
}
